Documents steps for component creation (#4982)
openshift/odo: website/docs/using-odo/create-component.md

sidebar_position: 1
---
# Creating components using odo

[Component](../getting-started/basics#component) is the most basic unit of operation for odo. And the way to create one is using `odo create` (short for `odo component create`) command.

> In simplest terms, when you "create" an odo component, you populate your current working directory with the file `devfile.yaml`. It is a manifest file that contains information about various resources (URL, Storage, Services, etc.) that correspond to your component, and will be created on the Kubernetes cluster when you execute `odo push` command. Most odo commands will first modify (add or remove configuration from) this file, and then subsequent `odo push` will create or delete the resources from the Kubernetes cluster. However, odo users are not expected to know how the `devfile.yaml` is organized; it is the odo commands that would create, update, or delete it.

Reviewer: >It is a manifest file that

I would rather prefer we say `Devfile is a manifest file` than calling it `It`.

Reviewer: >However, odo users are not expected to know how the `devfile.yaml` is organized; it is the odo commands that would create, update, or delete it.

What do you think about highlighting this line in a separate para perhaps or just highlighting it here? It would sort of take away the pressure of learning a new thing i.e. Devfile.

Author: Makes sense. Done both.

In this guide, we are going to create a Spring Boot component to deploy the [Spring Boot petclinic](https://github.com/spring-projects/spring-petclinic) project to a Kubernetes cluster.

Let's clone the project first:
```shell
git clone https://github.com/spring-projects/spring-petclinic
cd spring-petclinic
```

Next, let's create a project <!-- add link to project command reference here --> on the Kubernetes cluster in which we will be creating our component:
```shell
odo project create myproject
```
Alternatively, you could also use one of the existing projects on the cluster:
```shell
odo project list

> odo project set <project-name>  # replace <project-name> with a valid value from the list

Reviewer: Copying the whole block is not helpful. It would be better if we separate these 2 commands into different codeblocks.

```

odo supports interactive and non-interactive ways of creating a component.

> ## Interactively creating a component

Reviewer: This doesn't sound right, perhaps name it **_Creating a component interactively_** or **_The Interactive Way_** or **_Create a component: The interactive way_** ?

To interactively create a component, `cd` into the cloned petclinic project (already done if you copy-pasted the command above) and execute:
```shell
odo create
```
You will be prompted with a few questions one after the another. Let's go through each one of them to create a component.

1. First question is about selecting the component type:
    ```shell

>     odo create

Reviewer: How about adding `$`? Or would it not relate to a windows/mac user? 
```suggestion
    $ odo create
```

    ```
   By default, `java-maven` is selected for us. Since this is a Spring Boot application, we should be selecting `java-springboot`. 

    We can either scroll down to `java-springboot` using the arrow key, or start typing `spring` on the prompt. Typing `spring` will lead to odo filtering the component type based on your input.

2. Next, odo asks you to name the component:
    ```shell
    odo create                
    ? Which devfile component type do you wish to create java-springboot
    ? What do you wish to name the new devfile component (java-springboot) petclinic
    ```
    Let's name it `petclinic`.

3. Next, odo asks you for the project in which you would like to create the component. Use the project `myproject` that we created earlier or the one you had set using `odo project set` command
   ```shell
   odo create
   ? Which devfile component type do you wish to create java-springboot
   ? What do you wish to name the new devfile component java-springboot
   ? What project do you want the devfile component to be created in myproject
   ```
   Now you will have a `devfile.yaml` in your current working directory. But odo is just not done asking you questions yet.

> 4. Lastly, odo asks you if you would like to download a starter project. Since we already cloned the petclinic project, we answer in No by typing `n` and hitting the return key:

Reviewer: > Lastly, odo asks you if you would like to download a starter project.

Asking from a reader's perspective: What is a starter project?

I had this question a lot of times when I was getting started with odo; it might help to give `starter project` an alias, say `example project` in the brackets. Or better yet, add the hyperlink to the starter project section in the doc.

> hitting the return key

Return key? Shouldn't it be the enter key?

Author: > Asking from a reader's perspective: What is a starter project?
> 
> I had this question a lot of times when I was getting started with odo; it might help to give `starter project` an alias, say `example project` in the brackets. Or better yet, add the hyperlink to the starter project section in the doc.

Thanks for the catch! I had put a reference in the beginning of the doc. So I didn't put it here. But then I removed that statement and didn't remember to put it here. Updated. PTAL.

> Return key? Shouldn't it be the enter key?

They are the same. I have observed "Return" used more widely in docs.

   ```shell
   odo create
   ? Which devfile component type do you wish to create java-springboot
   ? What do you wish to name the new devfile component java-springboot
   ? What project do you want the devfile component to be created in myproject
   Devfile Object Validation
   ✓  Checking devfile existence [66186ns]
   ✓  Creating a devfile component from registry: stage [92202ns]
   Validation
   ✓  Validating if devfile name is correct [99609ns]
   ? Do you want to download a starter project (y/N) n
   ```
   
Your Spring Boot component is now ready for use.

> ## Non-interactively creating a component

Reviewer: Same as above, perhaps name it **_Creating a component non-interactively_** or **_The Non-Interactive Way_** or **_Create a component: The non-interactive way_**?

To non-interactively create the same component as we created in interactive mode, `cd` into the cloned petclinic project and execute:
```shell
odo create java-springboot petclinic -n myproject
```
Here `java-springboot` is the type of the component, `petclinic` is the name of the component, and `-n myproject` tells odo to use the project `myproject` for the mentioned `odo create` operation.

## Starter projects

> Besides creating a component for a code that wrote or cloned, you could also use "starter project" when creating a component.

Reviewer 2: ```suggestion
Besides creating a component for an existing code, you could also use "starter project" when creating a component.
```

Starter projects are example projects developed by the community to showcase the usability of devfiles. An odo user can use these starter projects by running `odo create` command in an empty directory.

### Starer projects in interactive mode

To interactively create a Java Spring Boot component using the starter project, you can follow the below steps:
```shell
mkdir myOdoComponent && cd myOdoComponent

> $ odo create

Reviewer: Would `&&` work on windows/mac? If not, should we separate the 2 commands?
Also, it would help with copying the command if you separate `odo create`  in a different code block.

```
In the questions that odo asks you next, provide answers like below:
```shell
$ odo create
? Which devfile component type do you wish to create java-springboot
? What do you wish to name the new devfile component myFirstComponent
? What project do you want the devfile component to be created in myproject
Devfile Object Validation
 ✓  Checking devfile existence [60122ns]
 ✓  Creating a devfile component from registry: stage [91411ns]
Validation
 ✓  Validating if devfile name is correct [35749ns]
? Do you want to download a starter project Yes

Starter Project
 ✓  Downloading starter project springbootproject from https://github.com/odo-devfiles/springboot-ex.git [716ms]

Please use `odo push` command to create the component with source deployed
```

### Starter projects in non-interactive mode

To non-interactively create a Java Spring Boot component using the starter project, you can follow the below steps:
```shell
mkdir myOdoComponent && cd myOdoComponent
odo create java-springboot myFirstComponent --starter
```

## Push the component to Kubernetes

> odo follows a "create & push" workflow for almost all the commands. Meaning, most odo commands won't create resources on Kubernetes cluster unless you run `odo push` command.

Reviewer: >most odo commands won't create resources on Kubernetes cluster

I am still confused whether we should mention just the K8s cluster or OCP as well.

Author: IMO k8s. OCP is for docs.openshift.com, not for upstream.

> Among the various ways described above, irrespective of how you created the component, the next step to create the resources for our component on the cluster would be to run `odo push`

Reviewer: Missing a period `.` at the end of the sentence.

> Note that first run of `odo push` could take a long time for the Spring Boot petclinic project. This is because of the dependencies being pulled by Spring Boot framework. Subsequent `odo push` runs shouldn't take longer when working on the same Kubernetes cluster.

Reviewer 2: Spring Boot framework is not pulling any dependencies. Spring boot is one of the dependency.
petclinic uses maven to manage java dependencies

Author: I have reworded it. PTAL.

> Note that first run of `odo push` could take a long time for the Spring Boot petclinic project. This is because of the dependencies being pulled by maven for Spring Boot framework. Subsequent `odo push` runs shouldn't take longer when working on the same Kubernetes cluster. 

Reviewer: > Subsequent `odo push` runs shouldn't take longer when working on the same Kubernetes cluster.

Again, should it just be K8s or OCP as well?

Author: https://github.com/openshift/odo/pull/4982#discussion_r686538322
